components/raftstore-v2/src/fsm/apply.rs

```diff
@@ -1,74 +1,102 @@
 
-use crate::{batch::ApplyContext, raft::Apply, router::ApplyTask};
+use crate::{
+    raft::Apply,
+    router::{ApplyRes, ApplyTask, PeerMsg},
+    tablet::CachedTablet,
+};
 
-pub struct ApplyFsm<EK: KvEngine> {
-    apply: Apply<EK>,
-    receiver: Receiver<ApplyTask>,
-    is_stopped: bool,
+/// A trait for reporting apply result.
+///
+/// Using a trait to make signiture simpler.
+pub trait ApplyResReporter {
+    fn report(&self, apply_res: ApplyRes);
 }
 
-impl<EK: KvEngine> ApplyFsm<EK> {
-    pub fn new(apply: Apply<EK>) -> (LooseBoundedSender<ApplyTask>, Box<Self>) {
-        let (tx, rx) = mpsc::loose_bounded(usize::MAX);
-        (
-            tx,
-            Box::new(Self {
-                apply,
-                receiver: rx,
-                is_stopped: false,
-            }),
-        )
-    }
-
-    /// Fetches messages to `apply_task_buf`. It will stop when the buffer
-    /// capacity is reached or there is no more pending messages.
-    ///
-    /// Returns how many messages are fetched.
-    pub fn recv(&mut self, apply_task_buf: &mut Vec<ApplyTask>) -> usize {
-        let l = apply_task_buf.len();
-        for i in l..apply_task_buf.capacity() {
-            match self.receiver.try_recv() {
-                Ok(msg) => apply_task_buf.push(msg),
-                Err(e) => {
-                    if let TryRecvError::Disconnected = e {
-                        self.is_stopped = true;
-                    }
-                    return i - l;
-                }
-            }
-        }
-        apply_task_buf.capacity() - l
+impl<F: Fsm<Message = PeerMsg>, S: FsmScheduler<Fsm = F>> ApplyResReporter for Mailbox<F, S> {
+    fn report(&self, apply_res: ApplyRes) {
+        // TODO: check shutdown.
+        self.force_send(PeerMsg::ApplyRes(apply_res)).unwrap();
     }
 }
 
-impl<EK: KvEngine> Fsm for ApplyFsm<EK> {
-    type Message = ApplyTask;
+/// Schedule task to `ApplyFsm`.
+pub struct ApplyScheduler {
+    sender: Sender<ApplyTask>,
+}
 
+impl ApplyScheduler {
     #[inline]
-    fn is_stopped(&self) -> bool {
-        self.is_stopped
+    pub fn send(&self, task: ApplyTask) {
+        // TODO: ignore error when shutting down.
+        self.sender.send(task).unwrap();
     }
 }
 
-pub struct ApplyFsmDelegate<'a, EK: KvEngine> {
-    fsm: &'a mut ApplyFsm<EK>,
-    apply_ctx: &'a mut ApplyContext,
+pub struct ApplyFsm<EK: KvEngine, R> {
+    apply: Apply<EK, R>,
+    receiver: Receiver<ApplyTask>,
 }
 
-impl<'a, EK: KvEngine> ApplyFsmDelegate<'a, EK> {
-    pub fn new(fsm: &'a mut ApplyFsm<EK>, apply_ctx: &'a mut ApplyContext) -> Self {
-        Self { fsm, apply_ctx }
+impl<EK: KvEngine, R> ApplyFsm<EK, R> {
+    pub fn new(
+        region_state: RegionLocalState,
+        res_reporter: R,
+        remote_tablet: CachedTablet<EK>,
+        logger: Logger,
+    ) -> (ApplyScheduler, Self) {
+        let (tx, rx) = future::unbounded(WakePolicy::Immediately);
```

Comment:
Is the future pool perf regression fixed?

Reply:
Not yet, but will.